docs/source/format/CDeviceDataInterface.rst

```diff
@@ -650,6 +652,331 @@ The stream source is not assumed to be thread-safe. Consumers wanting to
+
+    The ``const char*`` parameter exists for producers to provide any extra contextual information
+    they want. This is encoded in the same format as :c:member:`ArrowSchema.metadata`. If not ``NULL``,
+    the lifetime is only the scope of the call to this function. A consumer who wants to maintain
+    the additional metadata beyond the lifetime of this call *MUST* copy the value themselves.
+
+    A producer *MUST NOT* call this concurrently from multiple different threads.
+
+    The :c:member:`ArrowAsyncProducer.request` callback must be called to start receiving calls to this
+    handler.
+
+    This function *MUST* be able to be called re-entrantly on the same thread to allow for the
+    :c:member:`ArrowAsyncProducer.request` callback to potentially call this method recursively.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.on_error)(struct ArrowAsyncDeviceStreamHandler, int, const char*, const char*)
+
+    *Mandatory.* Handler to be called when an error is encountered by the producer. After calling
+    this, the ``release`` callback will be called as the last call on this struct. The parameters
+    are an ``errno``-compatible error code and an optional error message and metadata.
+
+    If the message and metadata are not ``NULL``, their lifetime is only valid during the scope
+    of this call. A consumer who wants to maintain these values past the return of this function
+    *MUST* copy the values themselves.
+
+    If the metadata parameter is not ``NULL``, to provide key-value error metadata, then it should
+    be encoded identically to the way that metadata is encoded in :c:member:`ArrowSchema.metadata`.
+
+    It is valid for this to be called by a producer with or without a preceding call to
+    :c:member:`ArrowAsyncProducer.request`. This callback *MUST NOT* call any methods of an
+    ``ArrowAsyncProducer`` object.
+
+.. c:member:: void (*ArrowAsyncDeviceStreamHandler.release)(struct ArrowAsyncDeviceStreamHandler*)
+
+    *Mandatory.* A pointer to a consumer-provided release callback for the handler.
+
+    It is valid for this to be called by a producer with or without a preceding call to
+    :c:member:`ArrowAsyncProducer.request`. This must not call any methods of an ``ArrowAsyncProducer``
+    object.
+
+.. c:member:: void* ArrowAsyncDeviceStreamHandler.private_data
+
+    *Optional.* An opaque pointer to consumer-provided private data.
+
+    Producers *MUST NOT* process this member. Lifetime of this member is handled by
+    the consumer, and especially by the release callback.
+
+The ArrowAsyncTask structure
+''''''''''''''''''''''''''''
+
+The purpose of using a Task object rather than passing the array directly to the ``on_next``
+callback is to allow for more complex and efficient thread handling. Utilizing a Task
+object allows for a producer to separate the "decoding" logic from the I/O, enabling a
+consumer to avoid transferring data between CPU cores (e.g. from one L1/L2 cache to another).
```

Comment:
Hmm, in real life the producer can have its own threads as well, since it's probably supposed to handle multiple in-flight consumer requests. Why would decoding be offloaded to consumer threads? This introduces a layer of complexity that will be difficult to make efficient.

Reply:
It's not required to offload the decoding to consumer threads, it's just a possible idea to reduce the cost of transferring data between CPU threads / improve cache handling as per the comment. (See @westonpace's comments for more info about the origin of this idea). For simplicity, most cases will likely perform the decoding on the producer side and then have `get_data` be nearly instantaneous to just return the data directly. 

This aspect merely makes it viable for use cases that require this complexity for efficiency to still be able to use this interface.

Reply:
But the consumer has zero knowledge whether it would be beneficial to offload `get_data` to separate threads, while the producer does. Why would this responsibility fall on the consumer?

Reply:
It's to allow a potential efficiency. This is adapted and based on the discussion that started [here](https://github.com/apache/arrow/pull/43632#issuecomment-2284669419) with @westonpace's comment.

But rather than the exact workflow that was proposed there I've adapted it to the following:

* Consumer calls `request` on the `ArrowAsyncProducer`
  * Producer schedules IO
  * When IO completes and data is available, Producer creates task to decode/decompress the data and calls `on_next_task` to hand it to the consumer
* Once consumer has a free thread it assigns it the task. Consumer runs task, decodes data (the decode logic is producer logic), and then processes the decoded data (consumer logic) all without a CPU core transfer (data stays in L2 cache).

In this way, the consumer determines when it has a free thread to process the data, it just potentially runs the producer logic to perform a decode/decompress before it processes the data. If this efficiency of managing the processor caches isn't necessary then the producer task could just return the data without any additional work if desired.

Reply:
As an existing real-world example, if the producer is reading from an IPC file, for each new RecordBatch it is able to decompress its buffers on multiple threads, which is good for latency.

Reply:
As long as this function is still not blocking on IO for the consumer, and the logic is managed by the producer, I don't see that as being an issue here. We're not dealing with multiple processes here as this is a C ABI, not a cross process interface, so ultimately it's just a matter of who is managing the thread pool. 

In the example you've provided, one of two things could happen:

1. The producer could decide that there's no benefit to the consumer to offload the decompression and simply perform that decompression itself before it calls `on_next_task`, so `get_data` is just handing off the `ArrowDeviceArray` object.
2. The producer decides there is benefit to allowing the consumer to determine *when* the decompression happens and bakes this logic into the `get_data` callback. The consumer chooses when to call `get_data` (a synchronous function call) which then utilizes the producer logic to perform the decompression (possibly across multiple producer threads if really desired) and then receives the final result.

In both cases, it allows for reduced pressure on the L2/L1 caches for handing the data off and the producer is making the decision as to whether or not there is benefit to offloading the logic. 

The documentation explicitly states that `get_data` is only allowed to be called **exactly once** on a given `ArrowAsyncTask` so there isn't any concern of calling it on a single task from multiple threads.

@lidavidm @westonpace @bkietz do any of you have any thoughts on this discussion? As this is a core aspect of the current design I'd like to get this hammered out before I officially send this to the mailing list for a vote (in case I need to explicitly change the design as a result of this discussion).

Reply:
You didn't answer my objection. In the IPC reader with buffer compression example, the producer could be able to leverage parallelism _for a single task_. But that's only if it manages threads by itself.

Reply:
If the producer wants to do the decompression then it can do so.  The "task" that gets delivered in this case is just a thin wrapper around the decoded data and doesn't do any work.

If the producer wants to leave decompression for the consumer then it can do so.  The "task" that gets delivered in this case actually does decompression.

Reply:
I think task structure is useful but we're not highlighting the real sweet spot for this flexibility. The reason not to do I/O and decoding on the same thread is that frequently we have separate thread pools for each, and we don't want to block for IO on a work thread nor do we want to do anything in an IO thread other than queue IO as immediately as we can. Any consumer who has distinct thread pools like this might be annoyed by a producer which insists on doing CPU bound work in a thread pool the consumer can't control. Tasks enable the producer to defer CPU bound work to the consumer

```suggestion
Using a Task object rather than simply passing the array directly to an ``on_next``
callback enables consumers to keep separate I/O and CPU work thread pools.
For example, a producer might construct Task objects after I/O is complete but
leave decoding logic until ``get_data`` is called. A consumer who knows this can
avoid blocking I/O threads by transferring Tasks to the CPU work thread pool.
After ``get_data`` completes the data will be processed immediately on the same
CPU work thread, avoiding transfer between CPU cores.
```

IIUC, Antoine's case of fanning out decoding is behavior that would most benefit a simpler consumer which has only one thread. That's also a useful case to support; what's the story for how a consumer would pass `use_threads=True` to a producer? Is that out of scope for this protocol, part of the factory interface which accepts a `ArrowAsyncDeviceStreamHandler*`?

Reply:
Note: all answers obviously biased by my current designs.

> The reason not to do I/O and decoding on the same thread is that frequently we have separate thread pools for each, and we don't want to block for IO on a work thread nor do we want to do anything in an IO thread other than queue IO as immediately as we can.

That is my goal.  In my current design the I/O thread pool belongs to the producer and the CPU thread pool belongs to the consumer.  Hence the need for the task interface.

> what's the story for how a consumer would pass use_threads=True to a producer?

What does this mean?  No I/O threads or no CPU threads?

Re I/O threads:

This is something negotiated outside the protocol.

Re CPU threads:

For the latter, in the "producer does not decode" case then the producer is not using CPU threads and so it's a non-issue.

In the "producer does decode" case then this would be something negotiated outside the protocol (e.g. as part of the call that starts producing).

In my current design the producer is created with a filesystem which has a "preferred parallelism" which is what controls the number of I/O threads.  The consumer controls the number of CPU threads by choosing how many tasks to run in parallel.

> Is that out of scope for this protocol

IMO, yes.

Reply:
I will admit there is one wrinkle I'm still ironing out :)

Decompression often happens at different scale than the consumers batch size.  For example, there might be 100K values per compression block but consumers might want to operate on batches of 32K.

This can force a thread transfer since three decode tasks share the same compression task and only one thread can do the work.  Still, this is something I would handle entirely in the consumer in my case (e.g. two CPU thread pools, one for decompression and one for decoding).

> a simpler consumer which has only one thread. That's also a useful case to support;

Simple cases are maybe more easily handled by the synchronous interface?

Reply:
> That is my goal. In my current design the I/O thread pool belongs to the producer and the CPU thread pool belongs to the consumer. Hence the need for the task interface.

If that's the goal for the spec then it should be clearly spelled out IMHO. _However_, it means that `get_data` probably shouldn't use any parallelism because of potential oversubscription. This should also be spelled out: `get_data` is a purely CPU, single-thread blocking task.

> If the producer wants to do the decompression then it can do so. The "task" that gets delivered in this case is just a thin wrapper around the decoded data and doesn't do any work.

That's a good point.

Reply:
> If that's the goal for the spec then it should be clearly spelled out IMHO. However, it means that get_data probably shouldn't use any parallelism because of potential oversubscription. This should also be spelled out: get_data is a purely CPU, single-thread blocking task.

It's a _possible_ goal but not _the_ goal.  For example, all my discussion so far has been "scanner is producer, query engine is consumer".  However, I think the "query engine is producer, analytics script is consumer" is a completely valid use case as well.  In the latter use case there is no separation of I/O and CPU to worry about, if both those thread pools exist then they both exist on the producer and I would full expect the tasks returned by the producer in this case to be shallow batch wrappers.

I agree that `get_data` should be a purely CPU single-threaded blocking task.

Reply:
It seems that we've got somewhat of a consensus here and I'll update the comments and docs accordingly as far as `get_data` being a purely CPU, single-threaded blocking task.

I'm fine with renaming it `extract_data` if others agree with @bkietz's suggestion. Thoughts?

Reply:
@pitrou any further comments here? Did we reach a consensus?